### PR TITLE
[SPARK-15754] [yarn] Not letting the credentials containing hdfs delegation tokens to be added in current user credential.

### PR DESCRIPTION
**Pull request by subrotosanyal**
## What changes were proposed in this pull request?

The credentials are not added to the credentials of UserGroupInformation.getCurrentUser(). Further if the client has possibility to login using keytab then the updateDelegationToken thread is not started on client.
## How was this patch tested?

ran dev/run-tests

### CODE REVIEW

**Comment by vanzin** [comment]
ok to test

**Comment by vanzin** [comment]
@subrotosanyal could you update the title to follow the convention? `[SPARK-xxxx][yarn] Summary.` Thanks!

**Comment by subrotosanyal** [comment]
@vanzin 
Updated the title.

**Comment by vanzin** [comment]
BTW the change looks fine, I just want to take a look at the rest of the code (the parts you're not touching) and also run some tests internally with kerberos enabled.

File: yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala [hunk 1]
```diff
@@ -21,9 +21,13 @@ import scala.collection.mutable.ArrayBuffer
 
 import org.apache.hadoop.yarn.api.records.YarnApplicationState
 
-import org.apache.spark.{SparkContext, SparkException}
-import org.apache.spark.deploy.yarn.{Client, ClientArguments, YarnSparkHadoopUtil}
+import org.apache.spark.SparkContext
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: These changes are unnecessary, and in fact we prefer the previous style.

File: yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala [hunk 2]
```diff
@@ -64,7 +68,7 @@ private[spark] class YarnClientSchedulerBackend(
     // SPARK-8851: In yarn-client mode, the AM still does the credentials refresh. The driver
     // reads the credentials from HDFS, just like the executors and updates its own credentials
     // cache.
-    if (conf.contains("spark.yarn.credentials.file")) {
+    if (!conf.contains(PRINCIPAL.key) && conf.contains("spark.yarn.credentials.file")) {
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: Since you're touching this, you could also use `CREDENTIALS_FILE_PATH`.

BTW I'm not entirely sure why this change is needed, or related to the issue you reported. You want to start the token renewer code exactly when `PRINCIPAL` is defined. You're pretty much disabling this code now, because `CREDENTIALS_FILE_PATH` is never set when `PRINCIPAL` is not.

Comment (subrotosanyal) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: hi @vanzin 
the method: org.apache.spark.deploy.yarn.ExecutorDelegationTokenUpdater.updateCredentialsIfRequired() invokes the same operation of updating the current user credential and correct me if I am wrong; if user has logged in through a keytab file already then ideally there shouldn't be any need to update the credentials at regular intervals. This thread of updating the credential might be useful for executors where there is only CREDENTAILS_FILE or delegation token.

Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnClientSchedulerBackend.scala]: That's a different issue, and I'm not sure it's really an issue, given what the comment above this line says.

This code is here for when Spark is managing your kerberos credentials and delegation tokens, i.e., when you pass `--principal` and `--keytab`. In that case, as the comment above says, the AM is tasked with renewing the credentials and tokens periodically, and this code exists so that the driver (which is not running in the AM in client mode) can get the new tokens.

You're basically breaking that feature by changing this. If your app is managing the kerberos login, you'd never pass `--principal` and `--keytab` (or the equivalent settings) to Spark, so you wouldn't run into this problem.

**Comment by subrotosanyal** [comment]
hi @vanzin 
Thanks for your comments.
I have removed the code change for YarnClientSchedulerBackend.
I will raise a separate ticket after further more analysis for the problem I am assuming to happen.

**Comment by vanzin** [comment]
Ran some kerberos tests here and all looks good, so I'm merging this (will also try 1.6.x if there are no conflicts). Thanks!
